vllm/multimodal/base.py

```diff
@@ -54,26 +52,32 @@ class MultiModalInputs(_MultiModalInputsBase):
     """
 
     @staticmethod
-    def _try_concat(tensors: List[NestedTensors]) -> BatchedTensors:
+    def _try_concat(nested_tensors: NestedTensors,
+                    is_nested: bool = False) -> NestedTensors:
         """
         If each input tensor in the batch has the same shape, return a single
         batched tensor; otherwise, return a list of :class:`NestedTensors` with
         one element per item in the batch.
         """
-        # may be list rather than tensors
-        if isinstance(tensors[0], list):
-            return [[t for t in tensor[0]]
-                    for tensor in cast(List[List[torch.Tensor]], tensors)]
-
-        tensors_ = cast(List[torch.Tensor], tensors)
-
-        unbatched_shape = tensors_[0].shape[1:]
-
-        for tensor in tensors_:
-            if tensor.shape[1:] != unbatched_shape:
-                return [tensor.squeeze(0) for tensor in tensors_]
-
-        return torch.cat(tensors_, dim=0)
+        if isinstance(nested_tensors, torch.Tensor):
+            # The input tensors already have a batch dimension; this will
+            # be re-stacked as long as there's a containing list that
+            # represents the batch dimension.
+            if is_nested:
+                return nested_tensors.squeeze(0)
+            else:
+                return nested_tensors
+
+        concatenated = [
+            MultiModalInputs._try_concat(t, is_nested=True)
+            for t in nested_tensors
+        ]
+        if any(isinstance(t, list) for t in concatenated):
+            return concatenated
+
+        tensors_ = cast(List[torch.Tensor], concatenated)
+        return torch.stack(tensors_) if all(t.shape == tensors_[0].shape
+                                            for t in tensors_) else tensors_
```

Comment:
Thanks for implementing my suggestion from the RFC! Would be great to add some tests for this.

Reply:
Thank you for laying out the plan!